text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
+
+# Alternatives
+
+- `impl` and `mut` restrictions could be attributes, similar to `#[non_exhaustive]`.
+
+# Prior art
+
+- The [`readonly` crate] simulates immutable fields outside of the defining module. Types with this
+  attribute cannot define `Deref`, which can be limiting. Additionally, it applies to all fields and
+  within the defining crate. The advantages of native read-only fields relating to borrow checking
+  also do not apply when using this crate.
+- The ability to restrict implementations of a trait can be simulated by a public trait in a private
+  module. This has the disadvantage that the trait is no longer nameable by external users,
+  preventing its use as a generic bound. Current diagnostics, while technically correct, are
+  unhelpful to downstream users.
+- Various other languages have read-only fields, including C++, C#, Java, TypeScript, Kotlin, and
+  Swift.
+
+# Unresolved questions
+
+- Should unrestricted "restrictions" be allowed to be explicitly stated, or should this be banned?
+  If allowed, should it be a warning?
+  - Both `impl` and `mut` make no sense to have unrestricted, as that is the default behavior. An
+    "unnecessary" restriction lint that is warn-by-default would be ideal. This would allow macros
+    to emit more flexible syntax. This warning could also be used for `pub(self)`.
+- How will restrictions work with `macro_rules!` matchers? There is currently a `vis` matcher, but
+  it is likely unwise to add a new matcher for each restriction. Should the new syntax be included
+  as part of the `vis` matcher? If so, it is important to note that restrictions are not the same
+  everywhere, if they are valid at all.
+  - Suggestions welcome!
+- What is the interaction between stability and restrictions?
+  - Suggestion: Visibility is an inherent part of the item; restrictions should be as well. Metadata
+    can be added in the future indicating when an item had its restriction lifted, if applicable.
+    The design for this is left to the language team as necessary.
+
+# Future possibilities
+
+- Sealed/exhaustive traits could happen in the future. This has the ability to impact coherence,
+  such that other crates could rely on the fact that the list of implementations is exhaustive. As
+  traits would default to unsealed, this does not have be decided now.
+- Trait items could gain proper visibility and/or restrictions of their own. This would allow
+  private and/or defaulted trait items that cannot be overridden.
+- Set-once fields could potentially occur in the future. Functionally, this would be "true"
+  read-only fields, in that they can be constructed but never mutated. They are not included in this
+  proposal as the use case is nor clear, nor is there an immediately obvious syntax to support this.
```

Comment:
```suggestion
  proposal as the use case is nor clear, nor is there an immediately obvious syntax to support this.
- We could change defaults in a future edition, such as to make `pub field: Type` be only mutable within the module rather than mutable everywhere. This seems unlikely, as this would be an incredibly disruptive change, and the benefits would have to be commensurately high.
```